server/src/main/java/io/crate/planner/optimizer/iterative/Memo.java

```diff
@@ -0,0 +1,225 @@
+ * license agreements.  See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.  Crate licenses
+ * this file to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.planner.optimizer.iterative;
+
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static java.util.Objects.requireNonNull;
+
+import io.crate.planner.operators.LogicalPlan;
+import io.crate.planner.operators.LogicalPlanVisitor;
+
+/**
+ * Stores a plan in a form that's efficient to mutate locally (i.e. without
+ * having to do full ancestor tree rewrites due to plan nodes being immutable).
+ * <p>
+ * Each node in a plan is placed in a group, and it's children are replaced with
+ * symbolic references to the corresponding groups.
+ * <p>
+ * For example, a plan like:
+ * <pre>
+ *    A -> B -> C -> D
+ *           \> E -> F
+ * </pre>
+ * would be stored as:
+ * <pre>
+ * root: G0
+ *
+ * G0 : { A -> G1 }
+ * G1 : { B -> [G2, G3] }
+ * G2 : { C -> G4 }
+ * G3 : { E -> G5 }
+ * G4 : { D }
+ * G5 : { F }
+ * </pre>
+ * Groups are reference-counted, and groups that become unreachable from the root
+ * due to mutations in a subtree get garbage-collected.
+ */
+public class Memo {
+    private static final int ROOT_GROUP_REF = 0;
+
+    private final int rootGroup;
+
+    private final Map<Integer, Group> groups = new HashMap<>();
+
+    private int nextGroupId = ROOT_GROUP_REF + 1;
+
+    public Memo(LogicalPlan plan) {
+        rootGroup = insertRecursive(plan);
+        groups.get(rootGroup).incomingReferences.add(ROOT_GROUP_REF);
+    }
+
+    public int getRootGroup() {
+        return rootGroup;
+    }
+
+    public LogicalPlan resolve(int group) {
+        return group(group).membership;
+    }
+
+    public LogicalPlan resolve(GroupReference groupReference) {
+        return resolve(groupReference.groupId());
+    }
+
+    public LogicalPlan extract() {
+        return extract(resolve(rootGroup));
+    }
+
+    private LogicalPlan extract(LogicalPlan node) {
+        return resolveGroupReferences(node, GroupReferenceResolver.from(this::resolve));
+    }
```

Comment:
What's the semantic difference between `extract` and `resolve` ?

Reply:
`resolve` is used for the usage for group reference resolver, `extract` is used to get the final Logical Plan once the optimization is finished.

Reply:
I meant more that it would be good to have some sort of explanation of what the methods do _without_ explaining them by referring _who_ uses them.

E.g it looks like `resolve(int group)` is accessing the plan  that's currently linked at group `group`.
`extract` is re-creating the full (sub) tree with new actual `LogicalPlan` nodes - without any group references?

It should be possible to gain this knowledge without looking at the implementation.